### PR TITLE
Part four query bus

### PR DESCRIPTION
**Pull request by ivanportillo**

### CODE REVIEW

File: src/Contexts/Shared/domain/CommandHandler.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import { Command } from './Command';
+
+export interface CommandHandler<T extends Command> {
+  subscribedTo(): Command;
+  handle(command: T): Promise<void>;
+}
```
Comment (JavierCane) [inline src/Contexts/Shared/domain/CommandHandler.ts:6]: 😍😊

File: src/Contexts/Shared/domain/CommandBus.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import { Command } from './Command';
+
+export interface CommandBus {
+  dispatch(command: Command): Promise<void>;
+}
```
Comment (JavierCane) [inline src/Contexts/Shared/domain/CommandBus.ts:5]: What do you think about enabling the [a TSLint `eofline` rule](https://palantir.github.io/tslint/rules/eofline/)?

[Right now it's disabled](https://github.com/CodelyTV/typescript-ddd-skeleton/blob/master/tslint.json#L7).

Also, we should add the `tslint` execution to [the `npm test`](https://github.com/CodelyTV/typescript-ddd-skeleton/blob/master/package.json#L15) or some other task to be executed in [the project GitHub Workflow](https://github.com/CodelyTV/typescript-ddd-skeleton/blob/master/.github/workflows/nodejs.yml#L24-L28).

Do you agree? Am I missing something?

File: src/Contexts/Shared/domain/QueryBus.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import { Query } from './Query';
+import { Response } from './Response';
+
+export interface QueryBus {
+  ask<R extends Response>(query: Query): Promise<R>;
+}
```
Comment (JavierCane) [inline src/Contexts/Shared/domain/QueryBus.ts:6]: 👌👌👌

File: src/Contexts/Shared/infrastructure/QueryBus/QueryHandlersInformation.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    return handlersMap;
+  }
+
+  public search(query: Query): QueryHandler<Query, Response> {
+    const queryHandler = this.queryHandlersMap.get(query.constructor);
+
+    if (!queryHandler) {
+      throw new QueryNotRegisteredError(query);
+    }
```
Comment (JavierCane) [inline src/Contexts/Shared/infrastructure/QueryBus/QueryHandlersInformation.ts:28]: Are we handling these exception somewhere?

File: src/Contexts/Shared/domain/CommandBus.ts (hunk 2 above)
Comment (ivanportillo) [reply src/Contexts/Shared/domain/CommandBus.ts:5]: Totally agree, I'm going to add it

File: src/Contexts/Shared/infrastructure/QueryBus/QueryHandlersInformation.ts (hunk 4 above)
Comment (ivanportillo) [reply src/Contexts/Shared/infrastructure/QueryBus/QueryHandlersInformation.ts:28]: This exception and the CommandNotRegisteredError are handled by the controller, throwing a 500 error. Is not handled as a specific error but its a general error.

https://github.com/CodelyTV/typescript-ddd-skeleton/blob/d6b7729e96f7ef64681101151e794f38913ed9d9/src/apps/mooc_backend/controllers/CoursesCounterGetController.ts#L21

**Pull request merged**